docs/madr/decisions/056-policy-matching-with-real-resources.md

```diff
@@ -0,0 +1,208 @@
+# Policy matching algorithm with real resources (MeshService, MeshExternalService, Mesh*Route)
+
+* Status: accepted
+
+## Context and Problem Statement
+
+We can think of the policy matching algorithm as a function
+that maps a set of policies provided by the user to the following structure:
+
+```yaml
+Rules:
+  - Subset:
+      - Key: kuma.io/service
+        Not: false
+        Value: foo
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 15s
+      idleTimeout: 20s
+    Origin: [ ... ]
+  - Subset:
+      - Key: kuma.io/service
+        Not: false
+        Value: bar
+    Conf:
+      connectionTimeout: 2s
+      http:
+        requestTimeout: 20s
+      idleTimeout: 20s
+    Origin: [ ... ]
+```
+
+This structure is used by both "from" and "to" policies plugins.
+Policy plugins use the `Rules` list to compute `Conf` for the known destination.
+In example, if we want to know the `Conf` for outbound with `kuma.io/service: bar` tags,
+we're going to iterate over `Rules` and check if our outbound belongs to the `Rules[].Subset`.
+
+The approach has a number of disadvantages:
+
+* even though it works well for "from" policies, using it for "to" policies doesn't feel natural
+  * there is no concept of "subsets" in outbounds, we have to artificially build `{kuma.io/service: $svc}` subset
+  * subset-based matching is build to handle subset overlapping scenarios, but it's never happening in outbounds
+* when referencing the real MeshHTTPRoute we had to use subset with artificial `__rule-matches-hash__` key
+* adding support for `MeshService` forces us to introduce another artificial key like `__resource-name__`
+* Inspect API with subsets is tricky when referencing real resources (like MeshHTTPRoute or MeshService)
+
+We need a new format that'd make things easier when it comes to referencing real resources in `to[]`.
+
+## Considered Options and Decision Outcome
+
+No changes for "from" policies, both policy matching algorithm and Inspect API stays the same.
+
+For "to" policies the algorithm is going to return the additional structure:
+
+```go
+type ToRules struct {
+    Rules         Rules 
+    ResourceRules map[UniqueResourceKey]ResourceRule // new field
+}
+
+// UniqueResourceKey is a way to uniquely identify a resource, should include ResourceType, Name and Mesh
+type UniqueResourceKey struct{}
+
+type ResourceRule struct {
+    Resource core_model.ResourceMeta
+    Conf     interface{}
+    Origin   []core_model.ResourceMeta
+}
+```
+
+For a simple targetRef with `name/namespace`:
+
+```yaml
+to:
+  - targetRef:
+      kind: MeshService
+      name: backend
+      namespace: finance
+    conf: conf1
+```
+
+we're going to produce a single `ResourceRule` entity:
+
+```yaml
+resourceRules:
+  - resource:
+      name: backend.finance
```

Comment:
One question, we were talking on Thursday meeting that you need 4 things to identify MeshService: `name`, `namespace`, `mesh` and `zone`. We have here name, namespace and mesh, what about zone? Does it even matter in this case?

Reply:
Zone will be encoded in name's hash. We might do it better with ResourceIdentity in the future, but it depends on the PoC